### ISSUE TITLE
Issues with deployment

### ISSUE THREAD

**Issue by tbranyen** [bocoup/email#8]
I gave it a shot, but am hung up on a few things.

1. Unable to get the correct key to deploy.
  - I ran `nest get-secrets` and `ssh-add config/secrets/*.pem` none of them seem to work.
  - Updated the ansible username to `tim` instead of `ubuntu`.
  - Changed the key location to `config/secrets/star.bocoup.com.pem`

  Got the following output:
  ```
  TASK: [deploy | make sure code is on the box and in sync] ********************* 
  failed: [email.bocoup.com -> 127.0.0.1] => {"cmd": "rsync --delay-updates -F --compress --archive --rsh 'ssh -i /home/tim/bocoup/email/config/secrets/star.bocoup.com.pem -S none -o StrictHostKeyChecking=no' --rsync-path=\"sudo rsync\" --include=package.json --include=bin/*** --include=server/*** --exclude='.*' --exclude='*' --delete-before --out-format='<<CHANGED>>%i %n%L' \"/home/tim/bocoup/email/\" \"tim@email.bocoup.com:/mnt\"", "failed": true, "rc": 12}
  msg: sudo: no tty present and no askpass program specified
  rsync: connection unexpectedly closed (0 bytes received so far) [sender]
  rsync error: error in rsync protocol data stream (code 12) at io.c(226) [sender=3.1.1]
  ```

2. How do I get the secrets into the deployment server?
  - I need the `config/secrets/ses.json` in production

3. The deploy setup only seems to run the deploy yaml file, how does provisioning happen on the target box?

``` yaml
---
# deploy code

- hosts: all
  gather_facts: false
  sudo: yes
  roles:
    - deploy
  handlers:
    - include: handlers/handlers.yml
```

**Comment by tbranyen** [comment]
Ping @tkellen not sure why I'm having so many issues, but I committed the ansible setup, would love your eyes tomorrow.

**Comment by tkellen** [comment]
We're moving to [vault](https://hashicorp.com/blog/vault.html) very soon, so copying secrets will be a little less wacky soon, but here is how I am doing it on the API:
https://github.com/bocoup/api/blob/master/deploy/ansible/roles/deploy/tasks/main.yml#L8-L15

...and here is how @iros is doing it on mobilevis:
https://github.com/bocoup/mobilevis/blob/ansible/deploy/ansible/roles/deploy/tasks/config.yml#L4-L11

We don't share the private key for boxes any longer--if someone leaves Bocoup and they have the private key we have no way of restricting access. At the moment, this means that Matt or I are spinning up new boxes and running [dploy](http://github.com/bocoup/dploy) on them manually. Once this has been done, you (and anyone else at Bocoup) can ssh (or deploy) directly to the machine without a `.pem`.

However, when you run ansible, you need to specify --ask-sudo-pass, or drop this in the deploy/ansible folder and run the command from the same directory to ensure it happens automatically:
https://github.com/bocoup/api/blob/master/deploy/ansible/ansible.cfg

If you haven't, it'd be worth it to take a minute to read this whole thing. @cowboy spent a ton of time on it and it's really well explained.
http://deployment-workflow.bocoup.com/

It's late and I'm headed to bed. Getting this stuff right the first time is tricky--I'll help ya get it sorted tomorrow if this doesn't get you unblocked now!

**Comment by cowboy** [comment]
@tkellen, you didn't even mention the [nest-cli](https://github.com/bocoup/nest-cli) tool or the in-progress [Deploying A Bocoup App](https://github.com/bocoup/infrastructure/wiki/Deploying-A-Bocoup-App) guide!

@tbranyen take a look at that stuff. We have a single command that does a lot of what you've had to do manually, it should be made easier to discover so that you don't have to do so much stuff manually!

**Comment by tbranyen** [comment]
Once that is done, I'll just need to scp the secrets and then wire up the code to read the correct path for SES credentials.

**Comment by tkellen** [comment]
Will take a look as soon as I get back--should be about 30 minutes.

**Comment by tkellen** [comment]
Have you touched the log file as the www-data user? I bet it doesn't have access to create it.

**Comment by tkellen** [comment]
this is how @jugglinmike did it:
https://github.com/bocoup/nest-weekly-review/blob/master/deploy/ansible/roles/configure/tasks/main.yml#L5-L6

...he also took the time to configure logrotate, which can be seen here (but is not a part of our deployment workflow):
https://github.com/bocoup/nest-weekly-review/blob/master/deploy/ansible/roles/configure/tasks/main.yml#L11-L12

**Comment by tbranyen** [comment]
After spending hours on the issue, I discovered a solution that fixed the problem [here](http://razius.com/articles/launching-services-after-vagrant-mount/).  For whatever reason on my system I need upstart to run at `vagrant -mounted` not `startup` as found in other's upstart scripts.  I still had to put in some host variable configuration logic to ensure that it runs on startup in production.  That post was written only for Vagrant development.

I'll move the documentation to finding out the answer to the README. Documenting here for now.

---

I would boot up the Vagrant box after provisioning and receive a 500 nginx page.  I could run the Node script just fine in the VM so something was up.  I modeled my upstart script after [@jugglinmike's weekly review one](https://github.com/bocoup/nest-weekly-review/blob/master/deploy/ansible/roles/services/templates/upstart.conf)

I discovered that upstart logs errors to /var/log under the service name, so the following blocks that show error output come from the following command:

```
vagrant@email:~$ sudo tail /var/log/upstart/app.log 
```

I looked at the error log and saw:

```
/bin/sh: 1: /bin/sh: cannot create /mnt/site/log: Read-only file system
```

Googling errors like this related to Vagrant were about symlinks, which wasn't the problem here.  I then Googled around for issues like this within upstart and found a post that said it might be [user permissions related](http://jeremykendall.net/2013/08/09/vagrant-synced-folders-permissions/) and that I should set the user id and group id in the `Vagrantfile`.

I checked the permissions on the log and they were incorrect:

```
-rw-r--r-- 1 vagrant vagrant   22 May  6 01:57 log
```

So I updated the Vagrantfile and re-provisioned. The permissions were fixed to `www-data`, but I got a new issue:

```
/proc/self/fd/9: 2: /proc/self/fd/9: cannot create /mnt/site/log: Permission denied
```

This was strange since I had changed the Vagrant mounts to `www-data` which was the user upstart was running as.  So I removed the log echo, but the server still wasn't working.  This time the error was:

```
    at node.js:814:3
module.js:338
    throw err;
          ^
Error: Cannot find module '/mnt/site/bin/index'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
```

Which was crazy, since the following ran just fine:

```
sudo -u www-data /usr/bin/node /mnt/site/bin/index
```

This made me think that permissions were still a problem.  So I Googled further and found this [superuser](http://superuser.com/a/471957) post that `setuid` and `setguid` were available in upstart 1.4 and up.  I had to look around to figure out how to determine the upstart version, which turns out to be [initctl version](http://askubuntu.com/a/554385).  I'm running a plenty new version at `1.12` so that's wasn't the problem.

Since the last error was related to being unable to find the file, I Google'd further and found the link that got it all working (linked at the top).  In development upstart queues off a vagrant-mounted ready event and in production it will queue off the startup event.  I don't know why this was required for me to get this working.  I made a host variable that changes from `vagrant-mounted` to `startup` in production.  The upstart file looks like:

```
start on {{upstart_startup}}
```

This was a very enlightening experience, but a huge time sink.

**Comment by tkellen** [comment]
:( :( :( i'm so sorry you had to get sucked into that black hole man.

I discovered it too awhile back. I wish I'd known you were hitting this so I could point you in the right direction:
https://github.com/bocoup/api/commit/b481731a4f40c73c14cf87db1877e066013ad8e9

The reason for this is that the vagrant share is mounted after the system finishes booting, which is well after upstart considers the system started.

One thing though... why are you logging to /mnt/site/<whatever> instead of /var/log/<whatever>?

T

**Comment by tbranyen** [comment]
No reason, I could switch to `/var/log`.  Should the filename be the FQDN? `/var/log/email.bocoup.com.log`

**Comment by tbranyen** [comment]
Well I didn't know I was having that issue either, I thought it was just permissions issues.  =/

**Comment by tbranyen** [comment]
Well I'm deploying now, fingers crossed.

**Comment by tkellen** [comment]
I'd probably use the project name for the log file, but whatever makes sense to you is a-ok. It's right there in the ansible config if someone can't find it.

**Comment by tkellen** [comment]
> Well I didn't know I was having that issue either, I thought it was just permissions issues. =/

:( blargh. computers.

**Comment by tbranyen** [comment]
Finally deployed!!!!!!!!!

http://email.bocoup.com/api/v1/send/

Boom!

I'll spend the rest of the week working through the inevitable bugs and locking it down.  I also want to get the job queue API written so I can create a nice frontend to monitor scheduled jobs.
